File: 0486-predict-the-winner/0486-predict-the-winner.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var PredictTheWinner = function(nums) {
    const n = nums.length
    const total = nums.reduce((acc, val) => acc + val, 0)
    const dp = {}
    const player1Score = helper(0, n-1)
    const player2Score = total - player1Score
    return player1Score >= player2Score
    
    function helper(l, r) {
        if ([l, r] in dp) return dp[[l, r]]
        if (l > r) return 0
        if (l === r) return nums[l]
        const score = Math.max(
            nums[l] + Math.min(helper(l+2, r), helper(l+1, r-1)),
            nums[r] + Math.min(helper(l, r-2), helper(l+1, r-1))
        )
        return dp[[l, r]] = score
    }
};

File: 0486-predict-the-winner/0486-predict-the-winner.py
class Solution:
    def predictTheWinner(self, nums: List[int]) -> bool:
        n = len(nums)
        dp = [[0] * n for _ in range(0, n)]
        for i in range(0, n):
            dp[i][i] = nums[i]
        for length in range(2, n+1):
            for i in range(0, n-length+1):
                j = i+length-1
                dp[i][j] = max(nums[i]-dp[i+1][j], nums[j]-dp[i][j-1])
        return dp[0][n-1] >= 0
        
        #
        
        @lru_cache(None)
        def rec(i: int, j: int) -> int:
            if i > j:
                return 0
            if i == j:
                return nums[i]
            left = nums[i] + min(rec(i+2, j), rec(i+1, j-1))
            right = nums[j] + min(rec(i+1, j-1), rec(i, j-2))
            return max(left, right)
        
        total = sum(nums)
        player1 = rec(0, len(nums)-1)
        player2 = total - player1
        return player1 >= player2
            
            